src/compiler/crystal/semantic/top_level_visitor.cr

```diff
@@ -564,6 +564,15 @@ module Crystal
             if default_value = member.default_value
               counter = interpret_enum_value(default_value, enum_base_type)
             end
+
+            if default_value.is_a?(Crystal::NumberLiteral)
+              enum_base_kind = enum_base_type.kind
+              node_kind = return_enum_kind(default_value)
+              if (enum_base_kind == :i32) && (enum_base_kind != node_kind)
```

Comment:
I just tried the change it and it works fine. The compiler knows `default_value` is a NumberLiteral because it's inside `if default_value.is_a?(Crystal::NumberLiteral)`. Maybe you didn't have that before?

In any case, doing it with a `return_enum_kind` should also give an error because you are not providing overloads for all ASTNode types (which means that the compiler knows that `default_value` can only be a `NumberLiteral`).